WordPress/Classes/Utility/DiskCache.swift

```diff
@@ -0,0 +1,273 @@
+import Foundation
+import CryptoKit
+
+/// An LRU disk cache that stores data in files on disk.
+///
+/// ``DiskCache`` uses LRU cleanup policy (least recently used items are removed
+/// first). The elements stored in the cache are automatically discarded if
+/// either *cost* or *count* limit is reached. The sweeps are performed periodically.
+///
+/// - important: It's possible to have more than one instance of ``DiskCache`` with
+/// the same path but it is not recommended.
+public actor DiskCache {
+    /// The path for the directory managed by the cache.
+    public nonisolated let rootURL: URL
+
+    /// The cache configuration.
+    public nonisolated let configuration: Configuration
+
+    public struct Configuration {
+        /// Size limit in bytes. `100 Mb` by default.
+        ///
+        /// Changes to the size limit will take effect when the next LRU sweep is run.
+        public var sizeLimit: Int = 1024 * 1024 * 100
+
+        /// When performing a sweep, the cache will remote entries until the size of
+        /// the remaining items is lower than or equal to `sizeLimit * trimRatio` and
+        /// the total count is lower than or equal to `countLimit * trimRatio`. `0.7`
+        /// by default.
+        var trimRatio = 0.7
+
+        /// The number of seconds between each LRU sweep. 30 by default.
+        /// The first sweep is performed right after the cache is initialized.
+        ///
+        /// Sweeps are performed in a background and can be performed in parallel
+        /// with reading.
+        public var sweepInterval: TimeInterval = 30
+
+        /// The delay after which the initial sweep is performed. 10 by default.
+        /// The initial sweep is performed after a delay to avoid competing with
+        /// other subsystems for the resources.
+        var initialSweepDelay: TimeInterval = 10
+
+        /// Initializes the configuration.
+        ///
+        /// - parameter sizeLimit: Size limit in bytes. `100 Mb` by default.
+        public init(sizeLimit: Int = 1024 * 1024 * 100) {
+            self.sizeLimit = sizeLimit
+        }
+    }
+
+    /// Creates a cache instance with a given `name`. The cache creates a directory
+    /// with the given `name` in a `.cachesDirectory` in `.userDomainMask`.
+    public init(name: String, configuration: Configuration = .init()) {
+        let cachesURL = DiskCache.getCachesURL()
+        self.init(url: cachesURL.appendingPathComponent(name, isDirectory: true), configuration: configuration)
+    }
+
+    private static func getCachesURL() -> URL {
+        guard let cachesURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+            DDLogError("Failed to instantiate cache. Caches directory not available")
+            return URL(fileURLWithPath: "/dev/null") // This should never happen
```

Comment:
There is a [`URL.cachesDirectory`](https://developer.apple.com/documentation/foundation/url/3988453-cachesdirectory?language=_5) (iOS 16+) that returns a non-nill `URL`, but unfortunately I can't use it. 

What I could do is make `DiskCache/rootURL` optional and add `guard let rootURL else { return }` checks in the `DiskCache` itself. But It shouldn't force the user to deal with optional or errors on every call site.

Reply:
Does this look good?

```swift
    private static func getCachesURL() -> URL {
        if #available(iOS 16, *) {
            return URL.cachesDirectory
        } else {
            guard let cachesURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
                DDLogError("Failed to instantiate cache. Caches directory not available")
                return URL(fileURLWithPath: "/dev/null") // This should never happen
            }
            return cachesURL
        }
    }
```

Reply:
I like that we're using the iOS 16 API where possible, but silently discarding data placed in the cache will cause hard-to-reason-about bugs (even if it _should_ never happen).

Two options for things that should never happen:

1. Crash the app – `!` is fine for this, but we could also use `precondition` (or any similar `Never`-returning method) to give ourselves better error data.
2. Allow the caller to gracefully handle the error.

Adding code paths that allow the app to continue on in an invalid state is a big part of why we have so many quality issues these days. Let's not introduce more 🙂

Reply:
A program should never rely on the cache being available, so it should not lead to bugs.

> Crash the app – ! is fine for this

I'm not sure a disk failure is a strong enough reason to crash an app on launch 😃 

> have so many quality issues these days

That's more philosophical than anything else. There is also a school of thought that the system should be resilient and _never_ crash because it can lead to lost user progress or data. The highly visible issue recently was the crashes.